### DISCUSSION TITLE
Does it matter which WidgetRef I use?

### DISCUSSION THREAD

**Discussion by wujek-srujek** [rrousselGit/riverpod#639] [Q&A] [answered]
When programming in Flutter, sometimes it does matter which `BuildContext` is used, be it for UI optimization/limiting rebuilds or simply because it makes a difference what 'stuff' is available.
Is there anything of this kind that I need to keep in mind using `WidgetRef`? Can I use a `ref` of a top level `ConsumerWidget.build(context, ref)` from a deeply nested event handler, for example? Or is it the same instance all the time, and it just doesn't matter?

**Comment by rrousselGit** [comment, marked as answer]
Yes, it matters for "watch".
Consider always using the deepest "ref"

If you use an ancestor's "ref" with watch, then it is that ancestor which will rebuild when the provider changes
